text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+    HealthCheckGracePeriod() *Duration
+    CloudMapOptions()        CloudMapOptions
+    PropagateTags()          PropagatedTagSource
+    EnableECSManagedTags()   *bool
+    DeploymentController()   DeploymentController
+}
+
+type BaseServicePropsIface interface {
+    BaseServiceOptionsIface  // embeddeded interface
+    LaunchType()             LaunchType
+}
+
+type BaseServiceOptions struct {
+    cluster                ICluster
+    desiredCount           *int
+    serviceName            *string
+    maxHealthyPercent      *int
+    minHealthyPercent      *int
+    healthCheckGracePeriod *Duration
+    cloudMapOptions        CloudMapOptions
+    propagateTags          PropagatedTagSource
+    enableECSManagedTags   *bool
+    deploymentController   DeploymentController
+}
+
+type BaseServiceProps struct {
+    BaseServiceOptions  // embedded (anonymous) field
+    launchType          LaunchType
+}
+
+func (o BaseServiceOptions) Cluster() ICluster { return o.cluster }
+func (o BaseServiceOptions) DesiredCount() int { return o.desiredCount }
+
+// ... etc
+
+```
+
+This allows the embedding interface (here, `BaseServicePropsIFace`) to "inherit" all the methods defined in the embedded interface
+(`BaseServiceOptionsIFace`). Similarly, the embedding struct (`BaseServiceProps`) would have access to all the properties of the embedded struct
+(`BaseServiceOptions`) through having it as an anonymous field (See: [Go playground example](https://play.golang.org/p/XRpTH3LN6bx)).
+
+```go
+props := BaseServiceProps{
+    BaseServiceOptions{
+      serviceName: "myService",
+      ...
+    },
+    launchType: "EC2",
+}
+
+props.serviceName  // => "myService"
+```
+
+### Notes/Concerns
+
+* When fields in the Go struct are exported, the struct cannot have a method with the same name (e.g. exported field `Bar` , and method `Bar()`.
+   Using a `Bar()` getter on a private field bar is most consistent with idiomatic Go;  however, this may make code generation more difficult since it
+would potentially necessitate converting public properties on the source (JSII) class into an interface implemented by the custom struct (Go “class”).
+The *alternative* to this is to prefix the methods in the generated Go interface with Get, e.g. `GetBar()` to avoid the name collision.
+* Like the AWS Go SDK, we can use pointers to primitive types to simulate that fields are optional (i.e. allow null values rather than default “empty”
+  values for each type, which are not nullable)
+* Generated Go interfaces corresponding to a datatype interface would need a suffix, e.g. Iface,  in order to disambiguate it from the struct. This is
+  a bit verbose, and it may be worth considering switching the naming (i.e. adding a suffix to the struct instead) if the interface name is what will
+primarily be used by the customer.
+
+* The alternative to having to implement both a struct and interface is simply translating the datatype interface into a struct. This option has the
```

Comment:
This is only acceptable if structs are also structurally typed. That is, if I can pass a `B` where an `A` is expected:

```go
type A struct {
  Foo int;
}

// One of the 2 Bs will model an 'interface B extends A' type declaration, which we still need to be able to
// pass into methods that expect an A
type B1 struct {
  Foo int;
  Bar string;
}

type B2 struct {
  A
  Bar string;
}
```

Reply:
Right, this is not possible in Go as A, B1 and B2 would be considered different types:
https://play.golang.org/p/JFIEXeyjUBN

The idea was to just use the struct instead of the interface as an argument, but that doesn't seem possible.

Reply:
Right.